### PR TITLE
dealii: add scalapack and various cleanups

### PR DESCRIPTION
**Pull request by davydden**
build tested on `macOS + clang + netlib-scalapack` and `Ubuntu + gcc + intel-mkl`.

ps. most changes are from `"` to `'`.

p.s. FYI @BenBrands

### CODE REVIEW

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 1]
```diff
@@ -309,36 +314,12 @@ def cmake_args(self):
...
+            netcdf = spec['netcdf-cxx'].libs + spec['netcdf'].libs
             options.extend([
                 '-DNETCDF_FOUND=true',
-                '-DNETCDF_LIBRARIES=%s;%s' % (
-                    join_path(netcdfcxx_lib_dir,
-                              'libnetcdf_c++.%s' % dso_suffix),
-                    join_path(netcdf_lib_dir,
-                              'libnetcdf.%s' % dso_suffix)),
+                '-DNETCDF_LIBRARIES=%s' % netcdf.joined(';'),
                 '-DNETCDF_INCLUDE_DIRS=%s;%s' % (
```
Comment (jppelteret) [inline var/spack/repos/builtin/packages/dealii/package.py]: Can you do the same joined path for the includes as you did for the libs?

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 2]
```diff
@@ -87,79 +88,83 @@ class Dealii(CMakePackage, CudaPackage):
...
-    depends_on("oce",              when='+oce')
-    depends_on("p4est",            when='+p4est+mpi')
-    depends_on("petsc+mpi~int64",  when='+petsc+mpi~int64')
-    depends_on("petsc+mpi+int64",  when='+petsc+mpi+int64')
-    depends_on("petsc@:3.6.4",     when='@:8.4.1+petsc+mpi')
+    depends_on('metis@5:+int64+real64',   when='+metis+int64')
+    depends_on('metis@5:~int64+real64',   when='+metis~int64')
+    depends_on('nanoflann',        when='@9.0:+nanoflann')
+    depends_on('netcdf+mpi',       when='+netcdf+mpi')
+    depends_on('netcdf-cxx',       when='+netcdf+mpi')
```
Comment (jppelteret) [inline var/spack/repos/builtin/packages/dealii/package.py]: I see from your modifications to the `netcdf-cxx` package thats its actually deprecated. Do we not support the newer bindings provided by `netcdf-cxx4` ?

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 3]
```diff
@@ -87,79 +88,83 @@ class Dealii(CMakePackage, CudaPackage):
...
+    depends_on('petsc+mpi+int64',  when='+petsc+mpi+int64')
+    depends_on('petsc@:3.6.4',     when='@:8.4.1+petsc+mpi')
     depends_on('python',           when='@8.5.0:+python')
-    depends_on("slepc",            when='+slepc+petsc+mpi')
-    depends_on("slepc@:3.6.3",     when='@:8.4.1+slepc+petsc+mpi')
-    depends_on("slepc~arpack",     when='+slepc+petsc+mpi+int64')
-    depends_on("sundials~pthread", when='@9.0:+sundials')
-    depends_on("trilinos+amesos+aztec+epetra+ifpack+ml+muelu+sacado+teuchos",       when='+trilinos+mpi~int64')
-    depends_on("trilinos+amesos+aztec+epetra+ifpack+ml+muelu+sacado+teuchos~hypre", when="+trilinos+mpi+int64")
+    depends_on('scalapack',        when='@9.0:+nanoflann')
```
Comment (jppelteret) [inline var/spack/repos/builtin/packages/dealii/package.py]: Oh, I suppose that this should be 
`depends_on('scalapack',        when='@9.0:+scalapack')`

Comment (davydden) [reply var/spack/repos/builtin/packages/dealii/package.py]: Absolutely, will fix

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 2 above)
Comment (davydden) [reply var/spack/repos/builtin/packages/dealii/package.py]: I was curious as well, let me try to build...

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 4]
```diff
@@ -87,79 +88,83 @@ class Dealii(CMakePackage, CudaPackage):
...
+    depends_on('slepc',            when='+slepc+petsc+mpi')
+    depends_on('slepc@:3.6.3',     when='@:8.4.1+slepc+petsc+mpi')
+    depends_on('slepc~arpack',     when='+slepc+petsc+mpi+int64')
+    depends_on('sundials~pthread', when='@9.0:+sundials')
+    depends_on('trilinos+amesos+aztec+epetra+ifpack+ml+muelu+sacado+teuchos',       when='+trilinos+mpi~int64')
```
[the remark this answers is not in the document]
Comment (davydden) [reply var/spack/repos/builtin/packages/dealii/package.py]: Yes, agree. Will add ROL here without 9.0

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 1 above)
Comment (davydden) [reply var/spack/repos/builtin/packages/dealii/package.py]: I am not too familiar with “headers” interface, so I would leave it as-is

Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py]: Ok!

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 2 above)
Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py]: Thanks for investigating :-)

Comment (davydden) [reply var/spack/repos/builtin/packages/dealii/package.py]: according to https://github.com/dealii/dealii/blob/master/cmake/modules/FindNETCDF.cmake#L40 we are looking for old c++ lib (without `4` at the end), will try to build anyway...

Comment (davydden) [reply var/spack/repos/builtin/packages/dealii/package.py]: nope, we don't support it. We have 
```
#include <netcdfcpp.h>
```
in `grid_in.cc` and this file is not there in `netcdf-cxx4`.

**Comment by davydden** [comment]
@jppelteret I addressed your comments.

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 2 above)
Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py]: Ah, ok. Well, thanks for looking into this anyway!

**Pull request merged**